doc/api/async_hooks.md

```diff
@@ -0,0 +1,671 @@
+
+
+##### `destroy(id)`
+
+* `id` {Number}
+
+Called either when the class destructor is run or if the resource is manually
+marked as free. For core C++ classes that have a destructor the callback will
+fire during deconstruction. It is also called synchronously from the embedder
+API `emitDestroy()`.
+
+Some resources, such as `HTTPParser`, are not actually destructed but instead
+placed in an unused resource pool to be used later. For these `destroy()` will
+be called just before the resource is placed on the unused resource pool.
+
+**Note:** Some resources depend on GC for cleanup. So if a reference is made to
+the `resource` object passed to `init()` it's possible that `destroy()` is
+never called. Causing a memory leak in the application. Of course if you know
+the resource doesn't depend on GC then this isn't an issue.
+
+#### `async_hooks.currentId()`
+
+* Returns {Number}
+
+Return the id of the current execution context. Useful to track when something
+fires. For example:
+
+```js
+console.log(async_hooks.currentId());  // 1 - bootstrap
+fs.open(path, (err, fd) => {
+  console.log(async_hooks.currentId());  // 2 - open()
+}):
+```
+
+It is important to note that the id returned has to do with execution timing.
+Not causality (which is covered by `triggerId()`). For example:
+
+```js
+const server = net.createServer(function onconnection(conn) {
+  // Returns the id of the server, not of the new connection. Because the
+  // on connection callback runs in the execution scope of the server's
+  // MakeCallback().
+  async_hooks.currentId();
+
+}).listen(port, function onlistening() {
+  // Returns the id of a TickObject (i.e. process.nextTick()) because all
+  // callbacks passed to .listen() are wrapped in a nextTick().
+  async_hooks.currentId();
+});
+```
+
+#### `async_hooks.triggerId()`
+
+* Returns {Number}
+
+Return the id of the resource responsible for calling the callback that is
+currently being executed. For example:
+
+```js
+const server = net.createServer(conn => {
+  // Though the resource that caused (or triggered) this callback to
+  // be called was that of the new connection. Thus the return value
+  // of triggerId() is the id of "conn".
+  async_hooks.triggerId();
+
+}).listen(port, () => {
+  // Even though all callbacks passed to .listen() are wrapped in a nextTick()
+  // the callback itself exists because the call to the server's .listen()
+  // was made. So the return value would be the id of the server.
+  async_hooks.triggerId();
+});
+```
+
+## Embedder API
+
+Library developers that handle their own I/O will need to hook into the
+AsyncWrap API so that all the appropriate callbacks are called. To accommodate
+this a JavaScript API is provided.
+
+### `class AsyncEvent()`
+
+The class `AsyncEvent` was designed to be extended from for embedder's async
+resources. Using this users can easily trigger the lifetime events of their
+own resources.
```

Comment:
Need more details on that or maybe you or @trevnorris can suggest the addition here?